File: N2.5/N2.5.swift
import Foundation

func wordConverter(_ fourthBlank: String) -> [String] {
    var convertedWord = [String]()
    var blankWord = fourthBlank.sorted()
    convertedWord.append(String(blankWord))
    while let nextPermutation = blankWord.nextPermutation() {
        convertedWord.append(String(nextPermutation))
    }
    return convertedWord
}

func stringSplitter(_ enterString: String) -> [String] {
    var splittedString = [String]()
    var miniWord = ""
    for (index, char) in enterString.enumerated() {
        if char == " " {
            splittedString.append(miniWord)
            miniWord = ""
        } else if index == enterString.count - 1 {
            miniWord += String(char)
            splittedString.append(miniWord)
        } else {
            miniWord += String(char)
        }
    }
    return splittedString
}

func translate(_ wordToTranslate: String, dictionary: [String: String]) -> String {
    var translatedWord = ""
    for char in wordToTranslate {
        if let morseCode = dictionary[String(char)] {
            translatedWord += morseCode
        }
    }
    return translatedWord
}

func unicChecker(_ translatedAllCombinations: [String], rnComb: String) -> String {
    var scoreOfSames = 0
    for word in translatedAllCombinations {
        if rnComb == word {
            scoreOfSames += 1
        }
    }
    if scoreOfSames > 1 {
        return ""
    } else {
        return rnComb
    }
}

func main() {
    let dictionary: [String: String] = [
        "a": ".-", "b": "-...", "w": ".--", "g": "--.", "d": "-..",
        "e": ".", "v": "...-", "z": "--..", "i": "..", "j": ".---",
        "k": "-.-", "l": ".-..", "m": "--", "n": "-.", "o": "---",
        "p": ".--.", "r": ".-.", "s": "...", "t": "-", "u": "..-",
        "f": "..-.", "h": "....", "c": "-.-.", "q": "--.-", "y": "-.--", "x": "-..-"
    ]
    
    var checkOfGo = true
    while checkOfGo {
        print("Enter String - ", terminator: "")
        guard let enterString = readLine() else { return }
        
        let splittedString = stringSplitter(enterString)
        
        var boolCheckOfWordLength = true
        for word in splittedString {
            if word.count > 9 {
                boolCheckOfWordLength = false
                break
            }
        }
        
        var boolCheckOfStringLength = true
        if enterString.count > 100 || enterString.count < 1 {
            boolCheckOfStringLength = false
        }
        
        if !boolCheckOfStringLength || !boolCheckOfWordLength {
            print("Please, enter correct string")
        } else {
            var allCombinations = [String]()
            for word in splittedString {
                for combination in wordConverter(word) {
                    allCombinations.append(combination)
                }
            }
            
            var translatedAllCombinations = [String]()
            for wordToTranslate in allCombinations {
                translatedAllCombinations.append(translate(wordToTranslate, dictionary: dictionary))
            }
            
            var unicCombinations = Set<String>()
            for rnComb in translatedAllCombinations {
                let unicCombination = unicChecker(translatedAllCombinations, rnComb: rnComb)
                if unicCombination == "" {
                    continue
                } else {
                    unicCombinations.insert(unicCombination)
                }
            }
            
            print()
            print("All combinations and their Morze translation:")
            for (index, output) in allCombinations.enumerated() {
                print("\(output) = (\(translatedAllCombinations[index]))")
            }
            
            print("Unic words:")
            for output in unicCombinations {
                print("(\(output))", terminator: " ")
            }
            
            print()
            print("____________________________________________________")
            print("               Value of unic words - \(unicCombinations.count)")
        }
        checkOfGo = false
    }
}

extension Array where Element: Comparable {
    mutating func nextPermutation() -> [Element]? {
        var i = count - 2
        while i >= 0 && self[i] >= self[i + 1] {
            i -= 1
        }
        
        if i < 0 {
            return nil
        }
        
        var j = count - 1
        while self[j] <= self[i] {
            j -= 1
        }
        
        swapAt(i, j)
        reverse(from: i + 1)
        
        return self
    }
    
    private mutating func reverse(from index: Int) {
        var i = index
        var j = count - 1
        while i < j {
            swapAt(i, j)
            i += 1
            j -= 1
        }
    }
}

main()

File: N2.5/N2.5.kt

import java.util.*

fun wordConverter(fourthBlank: String): List<String> {
    val convertedWord = mutableListOf<String>()
    var blankWord = fourthBlank.toCharArray().sorted().joinToString("")
    convertedWord.add(blankWord)
    var nextPermutation = true
    while (nextPermutation) {
        val nextWord = nextPermutation(blankWord)
        if (nextWord != null) {
            convertedWord.add(nextWord)
            blankWord = nextWord
        } else {
            nextPermutation = false
        }
    }
    return convertedWord
}

fun nextPermutation(word: String): String? {
    val charArray = word.toCharArray()
    var i = charArray.size - 2
    while (i >= 0 && charArray[i] >= charArray[i + 1]) {
        i--
    }
    if (i < 0) {
        return null
    }
    var j = charArray.size - 1
    while (charArray[j] <= charArray[i]) {
        j--
    }
    charArray.swap(i, j)
    charArray.reverse(i + 1)
    return String(charArray)
}

fun stringSplitter(enterString: String): List<String> {
    val splittedString = mutableListOf<String>()
    var miniWord = ""
    for (i in enterString.indices) {
        if (enterString[i] == ' ') {
            splittedString.add(miniWord)
            miniWord = ""
        } else if (i == enterString.length - 1) {
            miniWord += enterString[i]
            splittedString.add(miniWord)
        } else {
            miniWord += enterString[i]
        }
    }
    return splittedString
}

fun translate(wordToTranslate: String, dictionary: Map<String, String>): String {
    val translatedWord = StringBuilder()
    for (char in wordToTranslate) {
        val charString = char.toString()
        translatedWord.append(dictionary[charString])
    }
    return translatedWord.toString()
}

fun unicChecker(translatedAllCombinations: List<String>, rnComb: String): String? {
    var scoreOfSames = 0
    for (word in translatedAllCombinations) {
        if (rnComb == word) {
            scoreOfSames++
        }
    }
    return if (scoreOfSames > 1) {
        null
    } else {
        rnComb
    }
}

fun main() {
    val dictionary = mapOf(
        "a" to ".-", "b" to "-...", "w" to ".--", "g" to "--.", "d" to "-..",
        "e" to ".", "v" to "...-", "z" to "--..", "i" to "..", "j" to ".---",
        "k" to "-.-", "l" to ".-..", "m" to "--", "n" to "-.", "o" to "---",
        "p" to ".--.", "r" to ".-.", "s" to "...", "t" to "-", "u" to "..-",
        "f" to "..-.", "h" to "....", "c" to "-.-.", "q" to "--.-", "y" to "-.--", "x" to "-..-"
    )

    val scanner = Scanner(System.`in`)
    print("Enter String - ")
    val enterString = scanner.nextLine()

    val splittedString = stringSplitter(enterString)

    var boolCheckOfWordLength = true
    for (thirdBlank in splittedString) {
        if (thirdBlank.length > 9) {
            boolCheckOfWordLength = false
            break
        }
    }

    var boolCheckOfStringLength = true
    if (enterString.length > 100 || enterString.length < 1) {
        boolCheckOfStringLength = false
    }

    if (!boolCheckOfStringLength || !boolCheckOfWordLength) {
        println("Please, enter correct string")
    } else {
        val allCombinations = mutableListOf<String>()
        for (fourthBlank in splittedString) {
            for (combination in wordConverter(fourthBlank)) {
                allCombinations.add(combination)
            }
        }

        val translatedAllCombinations = mutableListOf<String>()
        for (wordToTranslate in allCombinations) {
            translatedAllCombinations.add(translate(wordToTranslate, dictionary))
        }

        val unicCombinations = mutableSetOf<String>()
        for (rnComb in translatedAllCombinations) {
            val unicCombination = unicChecker(translatedAllCombinations, rnComb)
            if (unicCombination != null) {
                unicCombinations.add(unicCombination)
            }
        }

        println()
        println("All combinations and their Morze translation:")
        for (i in allCombinations.indices) {
            println("${allCombinations[i]} = (${translatedAllCombinations[i]})")
        }
        println("Unic words:")
        for (output in unicCombinations) {
            print("($output) ")
        }
        println()
        println("____________________________________________________")
        println("Value of unic words - ${unicCombinations.size}")
    }
}

fun CharArray.swap(i: Int, j: Int) {
    val temp = this[i]
    this[i] = this[j]
    this[j] = temp
}

fun CharArray.reverse(startIndex: Int) {
    var i = startIndex
    var j = this.size - 1
    while (i < j) {
        this.swap(i, j)
        i++
        j--
    }
}
